### PR TITLE
Enable 'ip6tables' by default, don't require 'experimental'.

### PR DESCRIPTION
**Pull request by robmry**
**- What I did**

Enable `--ip6tables` by-default, don't require `--experimental`.

Note that this enables NAT/masquerading by default for IPv6 network. This is likely to remain the default, as it's a lowest-common-denominator that'll ensure bridge networks get IPv6 connectivity on an IPv6 host. In a follow-up PR, I'll make it possible to enable/disable masquerading separately for IPv4/IPv6.

With IPv6 masquerading disabled, running ip6tables by default means containers on bridge networks with routed IPv6 addresses only expose explicitly exposed/forwarded ports.

**- How I did it**

A series of commits - it's probably easiest to review them separately.

**- How to verify it**

Existing and new integration tests.

**- Description for the changelog**
```markdown changelog
Enable `ip6tables` by default, it is no longer `--experimental`.
```

### CODE REVIEW

File: libnetwork/drivers/bridge/setup_bridgenetfiltering.go [hunk 1]
```diff
@@ -12,27 +12,44 @@ import (
...
-// getIPVersion gets the IP version in use ( [ipv4], [ipv6] or [ipv4 and ipv6] )
-func getIPVersion(config *networkConfiguration) ipVersion {
-	ipVer := ipv4
-	if config.AddressIPv6 != nil || config.EnableIPv6 {
-		ipVer |= ipv6
+// setupIPv6BridgeNetFiltering checks whether IPv6 forwarding is enabled for the
+// bridge and, if it is, sets kernel param "bridge-nf-call-ip6tables=1" so that
+// packets traversing the bridge are filtered.
+func setupIPv6BridgeNetFiltering(config *networkConfiguration, _ *bridgeInterface) error {
+	if !config.EnableIPv6 || config.AddressIPv6 == nil {
```
Comment (akerouanton) [inline libnetwork/drivers/bridge/setup_bridgenetfiltering.go:32]: I'd rather not check if `AddressIPv6` is nil, thus would make the code more future proof.

File: integration/networking/bridge_test.go [hunk 2]
```diff
@@ -726,6 +727,69 @@ func TestNonIPv6Network(t *testing.T) {
...
+			defer c.ContainerRemove(ctx, id, containertypes.RemoveOptions{Force: true})
+
+			res, err := exec.Command("/usr/sbin/ip6tables-save").CombinedOutput()
+			assert.NilError(t, err)
+			if tc.expIPTables {
+				assert.Check(t, is.Contains(string(res), subnet))
+				assert.Check(t, is.Contains(string(res), bridgeName))
+			} else {
+				assert.Check(t, !strings.Contains(string(res), subnet),
+					fmt.Sprintf("Didn't expect to find '%s' in '%s'", subnet, string(res)))
```
Comment (akerouanton) [inline integration/networking/bridge_test.go:783]: `res` is going to be a multiline string, so this error won't be easy to read. I'd rather put `... in ip6tables-save output`.

File: daemon/daemon_unix.go [hunk 3]
```diff
@@ -737,11 +737,13 @@ func verifyDaemonSettings(conf *config.Config) error {
...
-	if !conf.BridgeConfig.EnableIPTables && !conf.BridgeConfig.InterContainerCommunication {
-		return fmt.Errorf("You specified --iptables=false with --icc=false. ICC=false uses iptables to function. Please set --icc or --iptables to true")
-	}
-	if conf.BridgeConfig.EnableIP6Tables && !conf.Experimental {
-		return fmt.Errorf("ip6tables rules are only available if experimental features are enabled")
+	if !conf.BridgeConfig.InterContainerCommunication {
+		if !conf.BridgeConfig.EnableIPTables {
+			return fmt.Errorf("You specified --iptables=false with --icc=false. ICC=false uses iptables to function. Please set --icc or --iptables to true")
+		}
+		if !conf.BridgeConfig.EnableIP6Tables {
```
Comment (akerouanton) [inline daemon/daemon_unix.go:744]: If you disable IPv6 at the kernel level and explicitly set `EnableIP6Tables=false` and `InterContainerCommunication=false`, then this check will fail. I think we should make this an error only if IPv6 has been enabled on the default bridge (ie. when `EnableIPv6` is true).

FWIW, we don't barf out if a custom network is created with icc=false, ipv6=true and ip6tables disabled. We should probably make config validation more consistent.

Comment (robmry) [reply daemon/daemon_unix.go:744]: Done.

(Is it possible to set `icc=false` for a custom network?)

File: integration/networking/bridge_test.go (hunk 2 above)
Comment (robmry) [reply integration/networking/bridge_test.go:783]: I temporarily added a `false &&` to make the test fail - the output looks like this ...

```
=== RUN   TestNoIP6Tables/ip6tables_off
    bridge_test.go:782: assertion failed: expression is false: false && !strings.Contains(string(res), subnet): Didn't expect to find 'fdb3:2511:e851:34a9::/64' in '# Generated by ip6tables-save v1.8.9 on Tue May  7 10:34:13 2024
        *nat
        :PREROUTING ACCEPT [0:0]
        :INPUT ACCEPT [0:0]
        :OUTPUT ACCEPT [0:0]
        :POSTROUTING ACCEPT [0:0]
        :DOCKER - [0:0]
        -A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
        -A OUTPUT ! -d ::1/128 -m addrtype --dst-type LOCAL -j DOCKER
        COMMIT
        # Completed on Tue May  7 10:34:13 2024
        # Generated by ip6tables-save v1.8.9 on Tue May  7 10:34:13 2024
        *filter
        :INPUT ACCEPT [0:0]
        :FORWARD DROP [0:0]
        :OUTPUT ACCEPT [2:232]
        :DOCKER - [0:0]
        :DOCKER-ISOLATION-STAGE-1 - [0:0]
        :DOCKER-ISOLATION-STAGE-2 - [0:0]
        :DOCKER-USER - [0:0]
        -A FORWARD -j DOCKER-USER
        -A FORWARD -j DOCKER-ISOLATION-STAGE-1
        -A DOCKER-ISOLATION-STAGE-1 -i docker0 ! -o docker0 -j DOCKER-ISOLATION-STAGE-2
        -A DOCKER-ISOLATION-STAGE-1 -j RETURN
        -A DOCKER-ISOLATION-STAGE-2 -o docker0 -j DROP
        -A DOCKER-ISOLATION-STAGE-2 -j RETURN
        -A DOCKER-USER -j RETURN
        COMMIT
        # Completed on Tue May  7 10:34:13 2024
        '
--- FAIL: TestNoIP6Tables (8.31s)
```

So, I think it's readable - and probably useful to see the full output.

File: libnetwork/drivers/bridge/setup_bridgenetfiltering.go (hunk 1 above)
Comment (robmry) [reply libnetwork/drivers/bridge/setup_bridgenetfiltering.go:32]: Done.

File: libnetwork/drivers/bridge/setup_bridgenetfiltering.go [hunk 4]
```diff
@@ -12,27 +12,44 @@ import (
...
+		return nil
+	}
+	bn := config.BridgeName
+	if bn == "" {
```
[the remark this answers is not in the document]
Comment (robmry) [reply libnetwork/drivers/bridge/setup_bridgenetfiltering.go:36]: Done.

File: integration/networking/bridge_test.go (hunk 2 above)
Comment (akerouanton) [reply integration/networking/bridge_test.go:783]: Okay, let's keep it as is then.

File: daemon/daemon_unix.go (hunk 3 above)
Comment (akerouanton) [reply daemon/daemon_unix.go]: > (Is it possible to set icc=false for a custom network?)

Yeah, you can create one with: `docker network create --opt=com.docker.network.bridge.enable_icc=false`

**Review by akerouanton — APPROVED** [review]

**Review by corhere — APPROVED** [review]

**Pull request merged**